tests/uri_router_test.py

```diff
@@ -9,6 +9,39 @@
 pytest.importorskip('srwl_bl')
 
 
+def test_error_for_bots():
+    from pykern import pkcompat
+    from pykern.pkcollections import PKDict
+    from pykern.pkunit import pkeq, pkexcept, pkok, pkre
+    from sirepo import http_request
+    from sirepo import srunit
+
+    fc = srunit.flask_client()
+    fc.sr_login_as_guest()
+
+    uri = '/get-application-data'
+    d = PKDict(simulationType='srw', method='NO SUCH METHOD')
+
+    # "Real" browsers get redirected to an error page with a 200 status
+    r = fc.post(
+        uri,
+        environ_base=PKDict(HTTP_USER_AGENT=http_request.USER_AGENT_MOZILLA),
+        json=d
+    )
+    pkeq(200, r.status_code)
+    pkok(
+        '/srw#/error' in pkcompat.from_bytes(r.data),
```

Comment:
Again the rule is parsing loose. Here we want to know that error appears, but we do not want to create an implicit coupling that `/srw` is a URL we expect. The implicit part is that applications begin with `/<app>`. While this is "known", the fewer tests that rely on this, the more easily it is changed. Therefore, you simply want to test for `/error` in the data, e.g.

```py
pkunit.pkre('/error', r.data)
```

If pkre doesn't support this, then we should fix pkre to support the conversion of the data. Again, parsing loose here: we can assume that pkre is going to do a reasonable job of matching the result. I noticed that it doesn't actually implement what it says it is implementing (action is type object). It should stringify its result like it does in other cases. So, in your never ending question please also fix https://github.com/radiasoft/pykern/issues/106

Reply:
1164018